modules/openid_connect/app/services/openid_connect/create_open_project_token.rb

```diff
@@ -0,0 +1,65 @@
+#-- copyright
+# OpenProject is a project management system.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2017 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+# +
+
+module OpenIDConnect
+  class CreateOpenProjectToken
+    def initialize(session)
+      @session = session
+    end
+
+    def call
+      if access_token.blank?
+        Rails.logger.error("Could not associate token to session: No access token")
+        return
+      end
+
+      user_session = find_user_session
+      if user_session.nil?
+        Rails.logger.error("Could not associate token to session: User session not found")
+        return
+      end
+
+      user_session.oidc_user_tokens.open_project.create!(access_token:, refresh_token:)
```

Comment:
TBD: Probably we should look at the actual audiences of the token instead of just assuming "open project" only.

Maybe the token already gives us more audiences.